cmake refactor

jbmertens opened the pull request:
is cmake fun now with codex??

  - CMake modularization: Introduced cmake/Options.cmake, Toolchain.cmake, Summary.cmake, and Features/* modules (GPU, IO, Math, Julia) to refactor and centralize configuration.
  - Find modules reorganized: Moved Find*.cmake files into cmake/Modules/; added new modules (e.g., FindGDAL.cmake, FindHighFive.cmake).
  - Top-level updates: Simplified CMakeLists.txt to include the new modules; adjusted lib/*/CMakeLists.txt to align with the refactor.
  - Feature detection UX: Added colorized feature summary and clearer -DUSE_<FEATURE>=ON|OFF controls; improved toolchain handling and flags.
  - Docs and README: Updated build instructions and compiling docs to reflect new CMake flow and feature summary behavior.
  
  
  TODO: 
 - [x] Do a quick pass and make sure we're listing all (kotekan-specific, non-cmake) options 
 - [x] TODO: fix OpenCL vs CUDA conflict, remove HIP.
 - [x] Fix BOOST vs DPDK conflict if it's easy?
 - [x] Then the default for most options will be auto-detect.
 - [x] Add a developer note to add new options both to the cmake summary, and docs, and version info.

jbmertens committed:
cmake refactor

jbmertens committed:
more cmake/docs

jbmertens committed:
cmake tweaks

jbmertens committed:
lint ci fix

jbmertens committed:
linting

jbmertens committed:
Merge branch 'chord' of github.com:kotekan/kotekan into jbm/cmake

jbmertens committed:
more explicit options and ui

jbmertens committed:
update version info

jbmertens committed:
linting

jbmertens committed:
some test fixes

jbmertens committed:
update workflows

jbmertens committed:
scheduled_tasks fix

jbmertens committed:
update sched tasks

jbmertens committed:
increase timeout

jbmertens committed:
update opencl

jbmertens committed:
lint

andrerenard commented on CMakeLists.txt:
```diff
@@ -4,315 +4,48 @@ project(
-message("GPU Modules Included: " ${GPU_MODULES})
-
-set(INPUT_MODULES "")
 if(${USE_AIRSPY})
```
This could maybe be moved to a "FeatureDigitizers" feature section.  I could see Keith adding more of these software defined radio systems in the future.

jbmertens replied:
Agreed, but I may wait until those are added before giving them a section.

andrerenard commented on README.md:
```diff
@@ -26,27 +27,22 @@ Cmake build options:
-* `-DUSE_OLD_DPDK=ON` - Include DPDK support for older (<19.11) versions (newer is included by default if installed).
-  Optional `-DRTE_SDK=<build-location>` and `-DRTE_TARGET=x86_64-native-linuxapp-gcc`
-  can be provided for non standard build locations.
+* `-DUSE_DPDK=ON|OFF|NEW|OLD` - Control DPDK support. Default `ON` auto‑detects: prefer NEW (>=19.11 via pkg‑config), else fall back to OLD (via FindDPDK). Use `OLD` for legacy (<19.11) installs (optionally with `-DRTE_SDK=<build-location>` and `-DRTE_TARGET=x86_64-native-linuxapp-gcc`). `OFF` disables DPDK entirely. The legacy flag `-DUSE_OLD_DPDK=ON` is still accepted and maps to `-DUSE_DPDK=OL ...
```
It's maybe out of the scope of this PR, but I'm dropping support for the "old" DPDK.  Also the optional `RTE_SDK` and  `RTE_TARGET` shouldn't be needed either, since even when I do custom DPDK installs, I just install them to standard locations.

jbmertens replied:
Sounds good, but yeah can we include this alongside dpdk changes?

andrerenard commented on README.md:
```diff
@@ -60,6 +56,8 @@ To build with CUDA:
 
+At the end of configuration, CMake prints a colorized feature summary indicating which features were enabled (found) or disabled (missing/explicitly off). Each feature row shows its toggle flag, e.g. `CUDA: ON (found, toggle: -DUSE_CUDA=ON/OFF)`. Use `-D<OPTION>=ON|OFF` to include or exclude a feature present on your system.
+
 To install kotekan (only works on CentOS at the moment):
```
We have long ago dropped support for CentOS, so this could be updated.

andrerenard commented on cmake/Features/FeatureDPDK.cmake:
```diff
@@ -0,0 +1,105 @@
+endif()
+
+# If building Boost tests, always disable DPDK
+if(WITH_BOOST_TESTS)
```
Why would we disable DPDK with boost tests?

jbmertens replied:
There's a conflict in the includes, if we compile with both of them on.
I could try to sort it out since none of the boost tests actually rely on dpdk, but it's a bit more work, and not a new restriction,
https://github.com/kotekan/kotekan/blob/develop/lib/CMakeLists.txt#L75

andrerenard commented on docs/sphinx/compiling/centos.rst:
```diff
@@ -63,5 +63,5 @@ To include **FFTW** stages:
     cd <kotekan_dir>/build
-    cmake .. -DUSE_FFTW=ON
-
+    # FFTW enables automatically if installed; disable with -DUSE_FFTW=OFF
```
Could just delete the entire CentOS file.   I rally doubt we'll want to bring back CentOS (well or equivalent ALMA Linux support).

andrerenard commented on docs/sphinx/compiling/cmake_options.rst:
```diff
@@ -0,0 +1,72 @@
+****************
+CMake Options
```
This is a great reworking of the CMAKE options.   One thing I've been tempted to do for sometime is to make most of the main features auto detect based on what libraries are installed.   e.g. for CUDA, DPDK, BLAZE, etc.   And instead move to an exclusion model, where if you _don't_ want a feature, you have to disable it.   The exception to this would be testing and ccache related things.    I think in most cases auto detection for most features is a lot nicer for the user.

jbmertens replied:
Yeah - it's a tradeoff between including if found, vs. waiting to compile in things you might not need. I think the new "summary" at the end of the cmake configure process is sufficiently informative that I could go either way.

geoffryan replied:
I feel for ease of reproducibility having explicit opt in/out flags is better? Otherwise on a different machine missing a dependency kotekan would silently build a different version, and its another step to check it configured how you wanted.

jbmertens replied:
TODO: fix OpenCL vs CUDA conflict, remove HIP. 
Fix BOOST vs DPDK conflict if it's easy?
Then the default for most options will be auto-detect.

andrerenard commented on docs/sphinx/compiling/general.rst:
```diff
@@ -22,42 +22,102 @@ Compiling **Kotekan**
+
+* Compilers and build tools: ``build-essential``, ``cmake``, ``pkg-config``, ``make``, ``git``
+* Libraries: ``libevent-dev``, ``libssl-dev``, ``libyaml-cpp-dev``
+* Python: ``python3``, ``python3-dev`` (plus runtime packages; see below for testing/docs)
```
I think this also should to include the python yaml, jinja2, requests, tabulate, and futures packages.  Which are needed either by the command line config parser or some of the performance monitoring scripts.  Technically they aren't needed for daemon installs, controlled by coco, but someone reading this is likely to want to run by command line.

andrerenard commented on external/gputils/CMakeLists.txt:
```diff
@@ -47,8 +47,7 @@ message(STATUS "CUFILES: ${CUFILES}")
 
-# Set architecture flags
-# set_target_properties(gputils PROPERTIES CUDA_ARCHITECTURES "80;86;89;90")
+# Set architecture flags set_target_properties(gputils PROPERTIES CUDA_ARCHITECTURES "80;86;89;90")
```
```suggestion
# Set architecture flags
```
I'm not sure the end of this comment is useful?

jbmertens replied:
This is from #1276 , but fine with me to remove --

andrerenard commented on external/n2k/CMakeLists.txt:
```diff
@@ -81,8 +81,7 @@ target_link_libraries(n2k gputils)
 
-# Set architecture flags
-# set_target_properties(n2k PROPERTIES CUDA_ARCHITECTURES "80;86;89;90")
+# Set architecture flags set_target_properties(n2k PROPERTIES CUDA_ARCHITECTURES "80;86;89;90")
```
Same here.

andrerenard reviewed (changes requested):
This looks really great!  I have a few comments, some of which you mind want to consider out of scope of this PR, but if it's convenient we could maybe address them now?

jbmertens committed:
merge chord

jbmertens committed:
remove centos refs

jbmertens committed:
clean up comments

jbmertens committed:
update compiling docs some

jbmertens committed:
lint

geoffryan commented:
This may also be out of scope, but the list of options currently has to be duplicated in `Options.cmake`, `make_version.cmake.in`, and `version.c.in`.  Also some show up in `python/kotekan/runner.py`   Could the `version/` functionality grab the data from `Options.cmake` automatically?

jbmertens commented:
> This may also be out of scope, but the list of options currently has to be duplicated in `Options.cmake`, `make_version.cmake.in`, and `version.c.in`. Also some show up in `python/kotekan/runner.py` Could the `version/` functionality grab the data from `Options.cmake` automatically?

The options/"features" are also duplicated elsewhere. I looked into de-duplicating them, and it is possible, but it isn't nice... essentially using cmake as a scripting language to do something it wasn't designed to do. We could consider another build system like meson (https://mesonbuild.com/), but I'm also fine with the changes as-is.

geoffryan committed:
Updating GPU Simulator & testing for N2K (#1279)

Updated data generation & gpu simulator for n2k updates.

Misc fixes to pass verify_cuda_n2k test and ensure clean exit of kotekan:
- Updated cudaDeviceInterface registry so instance destructors are called on shutdown.
- Amended kotekan return code to return 0 when test passes.

---------

Co-authored-by: James Mertens <jbcm627@gmail.com>

jbmertens committed:
hdf5 plugin feature

jbmertens committed:
fix merge conflict

jbmertens committed:
misc. options changes

jbmertens committed:
touchups

andrerenard commented on CMakeLists.txt:
```diff
@@ -4,315 +4,66 @@ project(
 
 # Fix a problem with homebrew linking on MacOS >= 10.14 See
 # https://stackoverflow.com/questions/54068035
 if(APPLE)
```
Ah looks like we could also just `if(APPLE)`

andrerenard commented on CMakeLists.txt:
```diff
@@ -4,315 +4,66 @@ project(
-endif()
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -lm")
+find_package(Threads REQUIRED)
+# Flags and link options handled in Toolchain; OpenCL install handled in FeatureGPU
```
Some of these codex comments feel kind of out of place.  I wouldn't be looking for OpenCL here in the future...

jbmertens replied:
Yeah; I think these are breadcrumbs left from bits of code that codex moved to another file.

andrerenard commented on cmake/Modules/FindLAPACKE.cmake:
There are some blocks in the `LICENSE` file for some of these Find scripts.   It would be good to update them with the new path.

andrerenard commented on config/ci-tests/test_n2k_writeHDF5.yaml:
```diff
@@ -0,0 +1,119 @@
+#
+# verify_cuda_n2k.yaml
+#
+# Config to test Kendrick's cuda N2 tensor-core kernel.
```
Was it intended for this to be included here?  or perhaps we just need to rebase?

jbmertens replied:
See comment below -- github doesn't update PR diffs when the base branch changes.

andrerenard commented on docs/sphinx/compiling/dpdk.rst:
```diff
@@ -85,9 +78,9 @@ Add the following to ``/etc/rc.local``, and make sure ``rc.local`` is executable
-    insmod /opt/dpdk-stable-16.11.10/x86_64-native-linuxapp-gcc/kmod/igb_uio.ko
-    PCI_NIC_LIST=`/opt/dpdk-stable-16.11.10/tools/dpdk-devbind.py --status | grep X710 | cut -b 6-12 | tr '\n' ' '`
-    /opt/dpdk-stable-16.11.10/tools/dpdk-devbind.py --bind igb_uio $PCI_NIC_LIST
+    insmod /opt/dpdk/x86_64-native-linuxapp-gcc/kmod/igb_uio.ko
```
I need to update these DPDK docs...  This is correct for the existing CHIME install, but the procedure is different now...

andrerenard commented on docs/sphinx/compiling/general.rst:
```diff
@@ -161,39 +224,32 @@ Cmake build options
-    and `-DRTE_TARGET=x86_64-native-linuxapp-gcc`
-    Not needed for newer versions of DPDK on Ubuntu 22.04+.
-    See :ref:`dpdk` for more details.
+* ``-DUSE_DPDK=ON|OFF``
```
Should we break up this list of CMAKE options to match that table you sent in slack.  It was kind of nice to see the auto detect vs. optional broken out.

andrerenard commented on docs/sphinx/compiling/macos.rst:
```diff
@@ -33,7 +33,8 @@ To include **FFTW** stages:
     brew install fftw
     cd <kotekan_dir>/build
-    cmake .. -DUSE_FFTW=ON
+    # FFTW enables automatically if installed; disable with -DUSE_FFTW=OFF
```
This comment seems unnecessary .

andrerenard commented on docs/sphinx/compiling/macos.rst:
```diff
@@ -59,7 +60,8 @@ To enable **HDF5** writing and linear algebra tools install:
     cmake --install cmake-build
     popd
-    cmake -DUSE_HDF5=ON -DUSE_LAPACK=ON
+    # HDF5 and LAPACK/Blaze enable automatically if installed; disable with -DUSE_HDF5=OFF -DUSE_LAPACK_BLAZE=OFF
```
Same here. This comment seems unnecessary.

andrerenard commented on docs/sphinx/compiling/ubuntu.rst:
```diff
@@ -27,7 +27,8 @@ To include **FFTW** stages:
     sudo apt-get install fftw3
     cd <kotekan_dir>/build
-    cmake .. -DUSE_FFTW=ON
+    # FFTW enables automatically if installed; disable with -DUSE_FFTW=OFF
```
And here too.

andrerenard commented on lib/cuda/cudaDeviceInterface.cpp:
This got included by mistake, or just needs a rebase?

andrerenard commented on lib/dpdk/CMakeLists.txt:
```diff
@@ -4,17 +4,19 @@ find_package(NUMA REQUIRED)
-    find_package(DPDK REQUIRED)
-    message("DPDK include dir: ${DPDK_INCLUDE_DIR}")
-    target_include_directories(kotekan_dpdk SYSTEM PRIVATE ${DPDK_INCLUDE_DIR})
+    message(FATAL_ERROR "kotekan_dpdk requested but KOTEKAN_DPDK_MODE is not NEW")
```
Do we actually need that `NEW` distinction anymore?  It's really just version > 19.11

andrerenard reviewed (approved):
Some small comments.  Biggest note is that this PR seems to be including changes from another PR...  I'm not sure if that's intended, or if the branch just needs to be rebased against the latest `chord`.   Otherwise, at the risk of more feature creep, happy to merge.

jbmertens commented:
> Some small comments. Biggest note is that this PR seems to be including changes from another PR... I'm not sure if that's intended, or if the branch just needs to be rebased against the latest `chord`. Otherwise, at the risk of more feature creep, happy to merge.

Yes, it's an annoying feature of github... the diff won't update unless I start a new PR.

https://github.com/orgs/community/discussions/16351

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

jbmertens committed:
update LICENSE

jbmertens committed:
lint

jbmertens committed:
update docs

jbmertens committed:
remove NEW dpdk option

jbmertens committed:
clang tweak

jbmertens committed:
lint

jbmertens committed:
merge

jbmertens committed:
fix clang errors

jbmertens committed:
more clang fixes

jbmertens committed:
addl fixes

jbmertens committed:
rm .keep
